iterators.tex

```diff
@@ -623,31 +650,186 @@
+\range{i}{j} denotes a range. Let $N$ be the smallest number of
+applications of \tcode{++i} necessary to make \tcode{bool(i == j)} be
+\tcode{true}. \tcode{SizedIterator<I>()} is satisfied if and only if:
+
+\begin{itemize}
+\item If $N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{j - i} is well-defined and equals $N$.
+
+\item If $-N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{i - j} is well-defined and equals $-N$.
+\end{itemize}
+\end{itemdescr}
+
+\pnum
+The \tcode{disable_sized} type predicate provides a mechanism to enable
+use of iterator types that meet the syntactic requirements but do not
+in fact satisfy \tcode{SizedIterator} with the library.
 
 \pnum
-The \tcode{Sentinel} concept defines requirements for a type that
+\enternote A program that instantiates a library template that requires
+an \tcode{Iterator} with such an iterator type \tcode{I} is ill-formed
+with no diagnostic required unless \tcode{disable_sized<I>} evalutes to
+\tcode{true}~(\ref{structure.requirements}). \exitnote
+
+\enternote The \tcode{SizedIterator} concept is satisfied by
+\tcode{RandomAccessIterator}s~(\ref{iterators.random.access}) and by
+counted iterators~(\ref{counted.iterator}).\exitnote
+
+\ednote{This concept also gives us a way to demote the category of \tcode{move_iterator}s to Input
+while retaining the ability of \tcode{move_iterator} pairs to communicate the range's size to
+container constructors.}
+} %% color{newclr}
+
+\rSec2[iterators.sentinel]{Concept Sentinel}
+\pnum
+The \tcode{Sentinel} concept \oldtxt{defines requirements for a type that
 is an abstraction of the past-the-end iterator. Its values can be
-compared to an iterator for equality.
+compared to an iterator for equality.}\newtxt{specifies requirements on
+the types \tcode{S} and \tcode{I} of a sentinel and an iterator that
+denote a range.}
 
 \indexlibrary{\idxcode{Sentinel}}%
-\begin{codeblock}
-  template <class T, class I>
+\begin{itemdecl}
+  template <class @\oldtxt{T}\newtxt{S}@, class I>
   concept bool Sentinel() {
-    return Regular<T>() &&
+    return Regular<@\oldtxt{T}\newtxt{S}@>() &&
       Iterator<I>() &&
-      EqualityComparable<T, I>();
+      EqualityComparable<@\oldtxt{T}\newtxt{S}@, I>();
   }
-\end{codeblock}
+\end{itemdecl}
+
+{\color{newclr}
+\begin{itemdescr}
+Let \tcode{s} and \tcode{t} be values of type \tcode{S},
+and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
+denotes a range. Types \tcode{S} and \tcode{I} satisfy
+\tcode{Sentinel<S, I>()} if and only if:
+
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
+      \tcode{s != i} are well-defined.
+
+\item \tcode{bool(i == s)} if and only if \tcode{i} and
+      \tcode{s} denote the same element.
+
+\item If \range{i}{s} and \range{i}{t} denote the same range,
+      then \tcode{s == t} is well-defined and \tcode{s}
+      equals \tcode{t}.
```

Comment:
I think the disconnect here is that we're arguing opposite meanings. You keep saying that `==` being true doesn't imply two values are equal when those values are not in the domain of `==`, which I agree with unreservedly. What I take issue with is the notion that it's reasonable for two values that *are* in the domain of `==`, for which `==` returns true, to have observably different behavior in equality-preserving expressions. That would contradict the basis of our semantics.

The axiom that began this entire thread states:
```tex
If \range{i}{s} and \range{i}{t} denote the same range,
then \tcode{s == t} is well-defined and \tcode{s}
equals \tcode{t}.
```
This does not imply, but states directly that if `[i,s)` and `[i,t)` denote the same range then `s` and `t` cannot have observably different behavior in equality-preserving expressions. The direct consequence is that for any `j` such that `[j,s)` denotes a range, `[j,t)` must also denote the same range. If `j` happens to be the iterator before `i`, and `t` is satisfied by `j`, we have a contradiction. We cannot say that sentinels only have meaning in connection with a particular range when this axiom gives them meaning across ranges. The axiom effectively requires that the predicates of `s` and `t` both be well-defined and have the same value for every iterator with which either `s` or `t` denotes a range, if there is a single instance in the program of two identical ranges `[i,s)` and `[i,t)`.

I'm sure the softness of the terms "denotes a range" and "same range" are contributing to the cross-talk here. We need to nail them down.

> Given some range [first,last) where last is a sentinel, and two iterators i and j that are reachable from first, then i==last && j==last implies i==j. That is the essential property of sentinels. And those semantics come straight from the algorithms and how they use iterators, sentinels, and ==.

I agree that is the essential property of sentinels; I don't agree with the implication that we need this axiom to ensure that property holds. If anything, it requires "If `[first,last)` denotes a range either `first == last` or `[++first,last)` denotes a range" which is [roughly the definition I have for "denotes a range" in my axioms for sentinels-as-predicates](https://github.com/CaseyCarter/cmcstl2/compare/advance...sentinel_predicates#diff-a54d540f5abe266043bcfee21eac7e99R552).

It's a contradiction in terms for a stateful type whose `==` doesn't reflect that state to satisfy `EqualityComparable`. Either the state doesn't contribute to behavior in EP expressions, in which case it isn't really a stateful type, or the state does have effects observable to EP expressions and the type does not satisfy `EqualityComparable`.